File: apps/android/Components/NostrUser/NostrProfileName.kt
package org.nmp.registry

import androidx.compose.material3.LocalTextStyle
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import java.util.UUID

/**
 * Inline display-name text for a Nostr profile.
 *
 * Two rendering modes:
 *   • [NostrProfileName] `(profile = …)` — caller already holds a
 *     [ProfileWire] (static, no claiming). Renders `profile.display`.
 *   • [NostrProfileName] `(pubkey = …)` — *self-claiming*. The component owns
 *     claiming the kind:0 it needs: it claims the author's profile from the
 *     [NostrProfileHost] on composition, reads the resolved projection
 *     reactively, and releases on disposal. This mirrors [NostrAvatar]'s
 *     claim/release lifecycle exactly.
 *
 * Display always comes from a Rust-formatted source — `displayName` when the
 * kind:0 has resolved, else the Rust-truncated `npubShort` (never reformat in
 * Kotlin, never raw hex). In the self-claiming mode, until the host has any
 * profile for the pubkey the component renders nothing rather than synthesize a
 * Kotlin-side abbreviation.
 *
 * Depends on `compose/user-avatar` for [ProfileWire] and [NostrProfileHost].
 */
@Composable
fun NostrProfileName(
    profile: ProfileWire,
    modifier: Modifier = Modifier,
    style: TextStyle = LocalTextStyle.current.copy(fontWeight = FontWeight.SemiBold),
    color: Color = Color.Unspecified,
) {
    val label = profile.display
    Text(
        text = label,
        style = style,
        color = color,
        maxLines = 1,
        overflow = TextOverflow.Ellipsis,
        modifier = modifier.semantics { contentDescription = "Display name: $label" },
    )
}

/**
 * Self-claiming display-name text keyed by `pubkey`.
 *
 * Claims the author's kind:0 from [NostrProfileHost] on composition (via
 * [DisposableEffect], like [NostrAvatar]), reads the resolved [ProfileWire]
 * reactively, and releases on disposal. The presentation layer owns claiming —
 * no event triggers a kernel kind:0 fetch of the author.
 *
 * Renders nothing until the host resolves a profile for `pubkey`, so the
 * fallback is always the Rust-formatted `npubShort`, never a Kotlin-side hex
 * abbreviation.
 */
@Composable
fun NostrProfileName(
    pubkey: String,
    modifier: Modifier = Modifier,
    style: TextStyle = LocalTextStyle.current.copy(fontWeight = FontWeight.SemiBold),
    color: Color = Color.Unspecified,
    consumerId: String? = null,
) {
    val profileHost = LocalNostrProfileHost.current
    val resolvedConsumerId = remember(pubkey, consumerId) {
        consumerId ?: "nostr-profile-name.${UUID.randomUUID()}"
    }

    DisposableEffect(pubkey, resolvedConsumerId) {
        profileHost?.claimProfile(pubkey, resolvedConsumerId)
        onDispose {
            profileHost?.releaseProfile(pubkey, resolvedConsumerId)
        }
    }

    // Read reactively from the host: recomposes when the kind:0 resolves.
    val resolved = profileHost?.profileForPubkey(pubkey) ?: return
    NostrProfileName(
        profile = resolved,
        modifier = modifier,
        style = style,
        color = color,
    )
}

File: apps/ios/Components/NostrUser/NostrProfileName.swift
import SwiftUI

/// Inline display-name text for a Nostr profile.
///
/// Two construction modes:
///   • `NostrProfileName(profile:)` — caller already holds a `ProfileWire`.
///     Claims the profile's pubkey on mount so the name stays fresh (mirrors
///     `NostrAvatar`), reads `profile.display`.
///   • `NostrProfileName(pubkey:)` — *self-claiming*. The component owns the
///     responsibility of claiming the kind:0 it needs: on mount it claims the
///     profile from the `NostrProfileHost`, reads the resolved projection
///     reactively, and releases on disappear. This mirrors `NostrAvatar`'s
///     claim/release lifecycle exactly, so any standalone name render anywhere
///     triggers resolution (claim-only invariant — every author-displaying
///     surface must self-claim).
///
/// Display always comes from a Rust-formatted source — `displayName` when set,
/// else `npubShort` (always Rust-formatted — aim.md §6.9). Until the host has
/// any profile for the pubkey, the self-claiming variant renders nothing rather
/// than synthesize a Swift-side abbreviation.
///
/// Depends on `swiftui/user-avatar` for `ProfileWire` and `NostrProfileHost`.
public struct NostrProfileName: View {
    @Environment(\.nostrProfileHost) private var profileHost

    /// Static profile supplied directly by the caller. `nil` when constructed
    /// in the self-claiming `pubkey:` mode.
    private let staticProfile: ProfileWire?
    /// Pubkey to claim (also the static profile's pubkey in static mode).
    private let pubkey: String
    public var font: Font
    public var color: Color

    @State private var generatedConsumerID: String
    @State private var claimedPubkey: String?

    /// Static variant: render the supplied `ProfileWire`. Claims its pubkey on
    /// mount so the name stays fresh.
    public init(
        profile: ProfileWire,
        font: Font = .headline,
        color: Color = .primary
    ) {
        self.staticProfile = profile
        self.pubkey = profile.pubkey
        self.font = font
        self.color = color
        self._generatedConsumerID = State(
            initialValue: "nostr-profile-name.\(UUID().uuidString)")
        self._claimedPubkey = State(initialValue: nil)
    }

    /// Self-claiming variant: claim the kind:0 for `pubkey` from the host, read
    /// the resolved profile reactively, release on disappear.
    public init(
        pubkey: String,
        font: Font = .body,
        color: Color = .primary,
        consumerID: String? = nil
    ) {
        self.staticProfile = nil
        self.pubkey = pubkey
        self.font = font
        self.color = color
        self._generatedConsumerID = State(
            initialValue: consumerID ?? "nostr-profile-name.\(UUID().uuidString)")
        self._claimedPubkey = State(initialValue: nil)
    }

    public var body: some View {
        let resolved = staticProfile ?? profileHost?.profile(forPubkey: pubkey)
        return Group {
            if let resolved {
                label(for: resolved)
            } else {
                // No kind:0 yet, and no Rust-formatted npubShort available.
                // Render nothing rather than a Swift-side abbreviation.
                EmptyView()
            }
        }
        .task(id: pubkey) {
            await MainActor.run {
                if let claimedPubkey, claimedPubkey != pubkey {
                    profileHost?.releaseProfile(
                        pubkey: claimedPubkey, consumerID: generatedConsumerID)
                }
                claimedPubkey = pubkey
                profileHost?.claimProfile(pubkey: pubkey, consumerID: generatedConsumerID)
            }
        }
        .onDisappear {
            if let claimedPubkey {
                profileHost?.releaseProfile(
                    pubkey: claimedPubkey, consumerID: generatedConsumerID)
                self.claimedPubkey = nil
            }
        }
    }

    private func label(for profile: ProfileWire) -> some View {
        Text(profile.display)
            .font(font)
            .foregroundStyle(color)
            .lineLimit(1)
            .accessibilityLabel("Display name: \(profile.display)")
    }
}
